keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,2144 @@
+```
+
+Once the pod is terminated and removed from the node, it should not be re-scheduled on the node by
+the deamonset controller until the node maintenance is complete.
+
+### Kubelet: Graceful Node Shutdown
+
+The current Graceful Node Shutdown feature has a couple of drawbacks when compared to
+NodeMaintenance:
+- It is application agnostic as it only provides a static grace period before the shutdown based on
+  priority. This does not always give the application enough time to react and can lead to data
+  loss or application availability loss.
+- The DaemonSet pods may be running important services (critical priority) that should be available
+  even during part of the shutdown. The daemon set controller does not have the observability of
+  the kubelet shutdown procedure and cannot infer which DaemonSets should stop running. The
+  controller needs to know which DaemonSets should run on each node with which priorities and
+  reconcile accordingly.
+
+To support these use cases we could introduce a new configuration option to the kubelet called
+`preferNodeMaintenanceDuringGracefulShutdown`.
+
+This would result in the following behavior:
+
+When a shutdown is detected, the kubelet would create a NodeMaintenance object for that node.
+Then it would block the shutdown indefinitely, until all the pods are terminated. The kubelet
+could pass the priorities from the `shutdownGracePeriodByPodPriority` to the NodeMaintenance,
+just without the `shutdownGracePeriodSeconds`. This would give applications a chance to react and
+gracefully leave the node without a timeout. [Pod Selection](#pod-selection) would ensure that user
+workloads are terminated first and critical pods are terminated last. 
+
+By default, all user workloads will be asked to terminate at once. The Evacuation API ensures that
+an evacuator is selected or an eviction API is called. This should result in a fast start of a pod
+termination.NodeMaintenance could then be used even by spot instances.
+
+The NodeMaintenance object could survive kubelet restarts, and the kubelet would always know if the
+node is under shutdown (maintenance). The cluster admin would have to remove the NodeMaintenance
+object after the node restart to indicate that the node is healthy and can run pods again. Admins
+are expected to deal with the lifecycle of planned NodeMaintenances, so reacting to the unplanned
+one should not be a big issue.
+
+If there is no connection to the apiserver (apiserver down, network issues, etc.) and the
+NodeMaintenance object cannot be created, we would fall back to the original behavior of Graceful
+Node Shutdown feature. If the connection is restored, we would stop the Graceful Node Shutdown and
+proceed with the NodeMaintenance.
+
+The NodeMaintenance would ensure that all pods are removed. This also includes the DaemonSet and
+Static pods.
+
+### Kubelet: Static Pods
+
+Currently, there is no standard solution for terminating static pods. We can advertise what state
```

Comment:
```suggestion
Currently, there is no standard solution for terminating Static Pods. We can advertise what state
```

Reply:
“static pods” is fine, IMO.

Reply:
I went with the capitalisation, given that it's the name of a feature. If you think it's fine, I will then delete the suggestions to remove some of the noise (for which I apologise in advance).

Reply:
We [use UpperCamelCase for API kinds](https://kubernetes.io/docs/contribute/style/style-guide/#use-upper-camel-case-for-api-objects), but then write some kinds in lower case when we're talking about the thing the object represents more than the API object.

> When you are generally discussing an API object, use [sentence-style capitalization](https://docs.microsoft.com/en-us/style-guide/text-formatting/using-type/use-sentence-style-capitalization).

Because static pods don't appear in the API (their mirror pods do), and because “static Pod” jars more than “static pod”, I'd use all-lowercase.

Reply:
@sftim, nice! Good to know for the future reference. Appreciated. :+1: